File: cpp/counting.cpp
const ll MOD = 998244353;
ll p0w(ll x,ll e){
	ll p = 1;
	ll acc = x;
	while(e){
		if(e&1) p = p*acc%MOD;
		acc = acc*acc%MOD;
		e /= 2;
	}
	return p;
}

const int MAX = 1e6 + 1;
const array<ll,MAX> fact = []{
	array<ll,MAX> fact = {1};
	rep(i,1,MAX) fact[i] = fact[i-1]*i%MOD;
	return fact;
}();
const array<ll,MAX> anti = []{
	array<ll,MAX> anti;
	anti[MAX-1] = p0w(fact[MAX-1],MOD-2);
	for( int i = MAX-2; i >= 0; i-- ) anti[i] = anti[i+1]*(i+1)%MOD;
	return anti;
}();
ll cmb(ll n, ll k){ return 0 <= k and k <= n ? fact[n]*anti[k]%MOD*anti[n-k]%MOD : 0; }

File: python/counting.py
MOD = 998244353
MAX = 6000
fact = [1]*(MAX+1)
for i in range(1,MAX+1): fact[i] = fact[i-1]*i%MOD
anti = [1]*(MAX+1)
anti[-1] = pow(fact[-1],-1,MOD)
for i in range(MAX-1,-1,-1): anti[i] = anti[i+1]*(i+1)%MOD
def cmb(n,k): return anti[n-k]*anti[k]%MOD*fact[n]%MOD if 0 <= k <= n else 0

#cat = [cmb(2*n,n)-cmb(2*n,n-1) for n in range(MAX//2 + 1)]
cat = [fact[2*n]*(anti[n]**2 - anti[n-1]*anti[n+1])%MOD for n in range(MAX//2 + 1)]
cat[0] = 1
